### PR TITLE
DP-28274: Add Analytics NEW tab for authors when editing

### PR DESCRIPTION
**Pull request by arthurbaghdas**
**Description:**
Explain the technical implementation of the work done.

**Jira:** (Skip unless you are MA staff)
DP-28274

**To Test:**
- [ ] Add steps to test this feature

---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)

### CODE REVIEW

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+   *
+   * @param \Drupal\node\NodeInterface $node
+   *   The upcasted node object.
+   *
+   * @return array
+   *   The iframe render array or a no match message render array.
+   */
+  public function build(NodeInterface $node) {
+    $looker_studio_url = $this->state->get('mass_analytics.looker_studio_url', '') . '?params=%7B"nodeId":[node:nid]%7D';
+    $config_url = $this->token->replace($looker_studio_url, ['node' => $node]);
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: What about calling this $iframe_url? It gets value not from config but from state.

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+   * Access callback for the page.
+   *
+   * @param \Drupal\node\NodeInterface $node
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   Return true if the node is one of the list.
+   */
+  public function access(NodeInterface $node) {
+    $access = FALSE;
+    $node->bundle();
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Remove

File: docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildForm($form, $form_state);
+    $default = $this->state->get('mass_analytics.looker_studio_url', '');
+
+    $form['looker_studio_url'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Looker Studio URL'),
+      '#description' => $this->t('Add Looker Studio URL without query string, it will be added in the code.'),
+      '#default_value' => $default,
+    ];
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php]: Why do we get the url from both state and config? If we want to use environment overrides, the config system can do that.

File: docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $value = $form_state->getValue('looker_studio_url');
+    $this->state->set('mass_analytics.looker_studio_url', $value);
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php]: Here we have a configForm that persists data via State. If we want to use state for some reason, just use a simple form, not a config form.

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php (hunk 2 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Why?

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php (hunk 1 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Done

File: docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php (hunk 3 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php]: Changed to use config

File: docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php (hunk 4 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Form/ConfigForm.php]: Changed everything to be config

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php (hunk 2 above)
Comment (weitzman) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: This line gets the bundle and does nothing with it. Did you want to assign it to a variable? It looks redundant with the line below it.

Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Right, removed.

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+use Drupal\token\Token;
+
+/**
+ * Class MassAnalyticsController.
+ *
+ * @package Drupal\mass_analytics\Controller
+ */
+class MassAnalyticsController extends ControllerBase {
+
+  const SCOPE = [
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: The name SCOPE seems unusual to me. I would expect BUNDLES or TYPES.

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   Return true if the node is one of the list.
+   */
+  public function access(NodeInterface $node) {
+    $access = FALSE;
+    if (in_array($node->bundle(), self::SCOPE)) {
+      $access = TRUE;
+    }
+    return AccessResult::allowedIf($access);
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]:
```suggestion
    return AccessResult::allowedIf(in_array($node->bundle(), self::SCOPE));
```

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php (hunk 5 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Changed

File: conf/drupal/config/config_ignore.settings.yml [hunk 7]
```diff
@@ -2,4 +2,5 @@ _core:
   default_config_hash: UVH1aJ4b44UM-VdPVN7hNNuuVqfReJxwfVeDQH1Hvsk
 ignored_config_entities:
   - mass_utility.settings
+  - mass_analytics.settings
```
Comment (weitzman) [inline conf/drupal/config/config_ignore.settings.yml]: If your goal with ignoring is env specific overrides, then I suggest adding code to https://github.com/massgov/openmass/blob/develop/docroot/sites/default/settings.acquia.php#L100-L131 instead.

Comment (arthurbaghdas) [reply conf/drupal/config/config_ignore.settings.yml]: I don't think it's environment specific.

Comment (arthurbaghdas) [reply conf/drupal/config/config_ignore.settings.yml:5]: I don't think it's environment specific.

Comment (weitzman) [reply conf/drupal/config/config_ignore.settings.yml]: OK, whats the motivation for the config ignore?

Comment (arthurbaghdas) [reply conf/drupal/config/config_ignore.settings.yml]: As Joe mentioned in our discussion about this ticket, we don't want the URL to be publicly available, which will the case if we export it.

Comment (weitzman) [reply conf/drupal/config/config_ignore.settings.yml]: Aha. What we have done for this in the past is just put a placeholder URL in the config yml file and then do a config override in settings.php like https://github.com/massgov/openmass/blob/develop/docroot/sites/default/settings.php#L137-L139

Comment (arthurbaghdas) [reply conf/drupal/config/config_ignore.settings.yml]: With the current implementation of analytics we don't have anything like that. It is configured from the UI with route_iframes module, And I don't see a benefit for complicating this to that level as you mentioned was done in the past. We have mass_utility settings ignored with config_ignore for example. I don't see a benefit for changing approach for this requirement.

Comment (weitzman) [reply conf/drupal/config/config_ignore.settings.yml]: I see my suggestion as simpler, not more complicated. Getting rid of config ignore is a simplification. And I dont see how the config ignore approach will work anywhere but Prod. The value will get reset on every --db-refresh. Thats inconvenient on local dev, for example. You will have to get an URL from Prod every time you want to work on this feature.

Feel free to move this to In Progress and assign to me and I am happy to complete the ticket.

File: docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+   *   The upcasted node object.
+   *
+   * @return array
+   *   The iframe render array or a no match message render array.
+   */
+  public function build(NodeInterface $node) {
+    $config = $this->config('mass_analytics.settings');
+    if (!empty($config->get('looker_studio_url'))) {
+      $looker_studio_url = $config->get('looker_studio_url') . '?params=%7B"nodeId":[node:nid]%7D';
+      $iframe_url = $this->token->replace($looker_studio_url, ['node' => $node]);
```
Comment (weitzman) [inline docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Is there a need to use the token system here? If not we already have a $node-id() and can put that in the querystring directly, thus removing the injection of token service to this Controller.

Comment (arthurbaghdas) [reply docroot/modules/custom/mass_analytics/src/Controller/MassAnalyticsController.php]: Yes we can remove token and use Node id directly.

**Pull request merged**